Mochi/src/MochiMochiProxy.hpp

```diff
@@ -0,0 +1,145 @@
+     * For the proxy just return the name of the proxy.
+     */
+    string name()
+    {
+        return string("MochiMochiProxy");
+    }
+
+    /**
+     * Train/update the model with the given training input.
+     */
+    void train(string* pInput, int dim)
+    {
+        for(vector<pair<string, BinaryOMLCreator*>>::iterator it=m_pBomlCreatorVector->begin(); it!=m_pBomlCreatorVector->end(); ++it)
+        {
+            it->second->train(pInput, dim);
+        }
+
+        /* Log the training data. */
+        if(m_pPropParser->isTrainingDataLogEnabled() == 1)
+        {
+            logTrainingData(m_pPropParser->getInputParamNames(), pInput);
+        }
+        
+    }
+
+    /**
+     * Train/update the model with the given training input and save/serialize the model.
+     * Note that for this proxy function the path argument is the parent directory path rather than the model file path.
+     */
+    void trainAndSave(string* pInput, size_t dim, const string modelDirPath)
+    {
+        for(vector<pair<string, BinaryOMLCreator*>>::iterator it=m_pBomlCreatorVector->begin(); it!=m_pBomlCreatorVector->end(); ++it)
+        {
+            it->second->trainAndSave(pInput, dim, modelDirPath + "/" + it->second->name());
+        }
+
+        /* Log the training data. */
+        if(m_pPropParser->isTrainingDataLogEnabled() == 1)
+        {
+            logTrainingData(m_pPropParser->getInputParamNames(), pInput);
+        }
+    }
+
+    /**
+     * Infer/predict the label with the given input.
+     * Note that for this proxy function the return value is not the prediction result.
+     */
+    int infer(string* pInput, size_t dim)
+    {
+        /* This vector will contain the predictions made by the trained algorithms. */
+        vector<pair<string, int>> inferences;
+
+        for(vector<pair<string, BinaryOMLCreator*>>::iterator it=m_pBomlCreatorVector->begin(); it!=m_pBomlCreatorVector->end(); ++it)
+        {
+            // TODO: insert prediction result into map (or write them directly into file?)
+            int label = it->second->infer(pInput, dim);
+            inferences.push_back(pair<string, int>(it->second->name(), label));
+        }
+
+        /* Log the inference results. */
+        logInferenceResult(m_pPropParser->getInputParamNames(), pInput, &inferences);
+        
+        /**
+         * Multiple model predictions are invoked.
+         * It makes no sense for this Proxy function to return a single prediction result. 
+         */
+        return 0;
+    }
+
+    /**
+     * Load a saved/serialized model.
+     * Note that for this proxy function the path argument is the parent directory path rather than the model file path.
+     */
+    void load(const string modelDirPath)
+    {
+        for(vector<pair<string, BinaryOMLCreator*>>::iterator it=m_pBomlCreatorVector->begin(); it!=m_pBomlCreatorVector->end(); ++it)
+        {
+            // TODO: Handle missing file exception (just skip and log the error?).
```

Comment:
Let's not forget about this, we had issues with loading models before silently failing.

Reply:
No worries, it's planned for the upcoming milestone: https://github.com/georgeslabreche/opssat-orbitai/issues/46.